File: app/src/main/java/ua/com/amicablesoft/android/wr/dal/Repository.java
package ua.com.amicablesoft.android.wr.dal;

import android.util.Log;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import ua.com.amicablesoft.android.wr.dto.UserDto;
import ua.com.amicablesoft.android.wr.models.Competition;
import ua.com.amicablesoft.android.wr.models.Powerlifter;
import ua.com.amicablesoft.android.wr.models.User;

/**
 * Created by lapa on 30.09.16.
 */

public class Repository implements IRepository {

    private DatabaseReference firebaseDatabase;
    private FirebaseAuth firebaseAuth;

    @Override
    public void userExist(final User user, final LoadUserCallback loadUserCallback) {
        firebaseDatabase = FirebaseDatabase.getInstance().getReference();
        firebaseDatabase.child("users").child(user.getId())
                .addListenerForSingleValueEvent(new ValueEventListener() {
                    @Override
                    public void onDataChange(DataSnapshot dataSnapshot) {
                        loadUserCallback.found(dataSnapshot.exists());
                    }

                    @Override
                    public void onCancelled(DatabaseError databaseError) {
                        // TODO error
                        loadUserCallback.found(false);
                    }
                }
        );
    }

    @Override
    public void userSave(User user) {
        firebaseDatabase = FirebaseDatabase.getInstance().getReference();
        firebaseDatabase.child("users").child(user.getId()).setValue(new UserDto(user.getEmail()));
    }

    @Override
    public void writeNewPowerlifter(String name, String lastName) {
        DatabaseReference firebaseDatabase = getDatabaseReference();
        String userId = getUserId();
        String key = firebaseDatabase.child("users/" + userId + "/powerlifters").push().getKey();
        Powerlifter powerlifter = new Powerlifter();
        powerlifter.setName(name);
        powerlifter.setLastName(lastName);
        Map<String, Object> childUpdates = new HashMap<>();
        childUpdates.put("users/" + userId + "/powerlifters/" + key, powerlifter);
        firebaseDatabase.updateChildren(childUpdates);

    }

    @Override
    public void getPowerlifters(final LoadPowerliftersCallback loadPowerliftersCallback) {
        DatabaseReference firebaseDatabase = getDatabaseReference();
        String userId = getUserId();
        firebaseDatabase.child("users/" + userId + "/powerlifters").addListenerForSingleValueEvent(
                new ValueEventListener() {
                    @Override
                    public void onDataChange(DataSnapshot dataSnapshot) {
                        ArrayList<Powerlifter> powerlifters = new ArrayList<>();
                        for (DataSnapshot snapshot: dataSnapshot.getChildren()) {
                            Powerlifter powerlifter = snapshot.getValue(Powerlifter.class);
                            powerlifters.add(powerlifter);
                        }
                        loadPowerliftersCallback.onPowerliftersLoaded(powerlifters);
                    }

                    @Override
                    public void onCancelled(DatabaseError databaseError) {
                        Log.w("getLifters:onCancelled", databaseError.toException());
                        loadPowerliftersCallback.onDataNotAvailable();
                    }
                }
        );
    }

    @Override
    public void writeNewCompetition(String nameOfCompetition) {
        DatabaseReference firebaseDatabase = getDatabaseReference();
        String userId = getUserId();
        String key = firebaseDatabase.child("users/" + userId + "/competitions").push().getKey();
        Competition competition = new Competition();
        competition.setCompetition(nameOfCompetition);
        Map<String, Object> childUpdates = new HashMap<>();
        childUpdates.put("users/" + userId + "/competitions/" + key, competition);
        firebaseDatabase.updateChildren(childUpdates);
    }

    @Override
    public void getCompetitions(final LoadCompetitionsCallback loadCompetitionsCallback) {
        DatabaseReference firebaseDatabase = getDatabaseReference();
        String userId = getUserId();
        firebaseDatabase.child("users/" + userId + "/competitions").addListenerForSingleValueEvent(
                new ValueEventListener() {
                    @Override
                    public void onDataChange(DataSnapshot dataSnapshot) {
                        ArrayList<Competition> competitions = new ArrayList<>();
                        Competition c = new Competition();
                        c.setCompetition("- None -");
                        competitions.add(c);
                        for (DataSnapshot snapshot: dataSnapshot.getChildren()) {
                            Competition competition = snapshot.getValue(Competition.class);
                            competitions.add(competition);
                        }
                        loadCompetitionsCallback.onCompetitionsLoaded(competitions);
                    }

                    @Override
                    public void onCancelled(DatabaseError databaseError) {
                        Log.w("getCompets:onCancelled", databaseError.toException());
                        loadCompetitionsCallback.onDataNotAvailable();
                    }
                }
        );
    }

    private DatabaseReference getDatabaseReference() {
        firebaseAuth = FirebaseAuth.getInstance();
        return FirebaseDatabase.getInstance().getReference();
    }

    private String getUserId() {
        return firebaseAuth.getCurrentUser().getUid();
    }
}

File: app/src/main/java/ua/com/amicablesoft/android/wr/dal/Repository.kt
/*
 * Created by Olha Tymoshenko on 5/24/18 8:35 PM.
 * Copyright (c) 2018. All rights reserved.
 */

package ua.com.amicablesoft.android.wr.dal

import android.util.Log
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import ua.com.amicablesoft.android.wr.dto.UserDto
import ua.com.amicablesoft.android.wr.models.Competition
import ua.com.amicablesoft.android.wr.models.Powerlifter
import ua.com.amicablesoft.android.wr.models.User

class Repository : IRepository {

    override fun userExist(user: User, loadUserCallback: IRepository.LoadUserCallback) {
        val firebaseDatabase = FirebaseDatabase.getInstance().reference
        firebaseDatabase.child("users").child(user.id).addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onCancelled(p0: DatabaseError) {
                //TODO error
                loadUserCallback.found(false)
            }

            override fun onDataChange(p0: DataSnapshot) {
                loadUserCallback.found(p0.exists())
            }
        })
    }

    override fun userSave(user: User) {
        val firebaseDatabase = FirebaseDatabase.getInstance().reference
        firebaseDatabase.child("users").child(user.id).setValue(UserDto(user.email))
    }

    override fun writeNewPowerlifter(name: String, lastName: String) {
        val databaseReference = FirebaseDatabase.getInstance().reference
        val userId = FirebaseAuth.getInstance().currentUser?.uid
        val key = databaseReference.child("users/$userId/powerlifters").push().key
        val powerlifter = Powerlifter(name, lastName)
        val childUpdates = mapOf(Pair("users/$userId/powerlifters/$key", powerlifter))
        databaseReference.updateChildren(childUpdates)
    }

    override fun getPowerlifters(loadPowerliftersCallback: IRepository.LoadPowerliftersCallback) {
        val databaseReference = FirebaseDatabase.getInstance().reference
        val userId = FirebaseAuth.getInstance().currentUser?.uid
        databaseReference.child("users/$userId/powerlifters").addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onCancelled(p0: DatabaseError) {
                Log.w("getLifters:onCancelled", p0.toException())
                loadPowerliftersCallback.onDataNotAvailable()
            }

            override fun onDataChange(p0: DataSnapshot) {
                val powerlifters = p0.children.map { s -> s.getValue(Powerlifter::class.java) }
                loadPowerliftersCallback.onPowerliftersLoaded(powerlifters)
            }
        })
    }

    override fun writeNewCompetition(competition: String) {
        val databaseReference = FirebaseDatabase.getInstance().reference
        val userId = FirebaseAuth.getInstance().currentUser?.uid
        val key = databaseReference.child("users/$userId/competitions").push().key
        val competitionObject = Competition(competition)
        val childUpdates = mapOf(Pair("users/$userId/competitions/$key", competitionObject))
        databaseReference.updateChildren(childUpdates)
    }

    override fun getCompetitions(loadCompetitionsCallback: IRepository.LoadCompetitionsCallback) {
        val databaseReference = FirebaseDatabase.getInstance().reference
        val userId = FirebaseAuth.getInstance().currentUser?.uid
        databaseReference.child("users/$userId/competitions").addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onCancelled(p0: DatabaseError) {
                Log.w("getCompets:onCancelled", p0.toException())
                loadCompetitionsCallback.onDataNotAvailable()
            }

            override fun onDataChange(dataSnapshot: DataSnapshot) {
                val competitions = dataSnapshot.children.map { child ->
                    child.getValue(Competition::class.java)
                }

                val noneCompetition = Competition("- None -")

                val allCompetitions = listOf(noneCompetition).plus(competitions)
                loadCompetitionsCallback.onCompetitionsLoaded(allCompetitions)
            }
        })
    }
}